### ISSUE TITLE
Taint edges are discarded when `ArrayAccess::offsetGet()` is reached through `$x['k']` sugar

### ISSUE THREAD

**Issue by alies-dev** [vimeo/psalm#11912]
### Summary

Taint data flow is dropped when `ArrayAccess::offsetGet()` is reached through the `$x['k']` array-access sugar. The equivalent explicit `$x->offsetGet('k')` call is analyzed correctly.

This is a silent false negative, which is the worst failure mode for a security check: the analysis reports success while the flow it was asked to track has vanished. It affects every `ArrayAccess`-based API (Symfony parameter bags, Doctrine collections, Laravel collections, PSR-7 style containers, and any library that annotates its accessor), because `$x['k']` is the form essentially all callers write and `$x->offsetGet('k')` is the form almost nobody writes.

The loss is **not** limited to `@psalm-taint-source`. Any taint edge crossing the sugar is discarded, including a plain argument-to-return pass through. Both cases are shown below.

### Case 1: `@psalm-taint-source` on the accessor

```php
<?php
/** @implements ArrayAccess<string, mixed> */
final class Bag implements ArrayAccess
{
    /** @psalm-taint-source input */
    #[Override] public function offsetGet(mixed $offset): mixed { return $_GET['q'] ?? ''; }
    #[Override] public function offsetExists(mixed $offset): bool { return true; }
    #[Override] public function offsetSet(mixed $offset, mixed $value): void {}
    #[Override] public function offsetUnset(mixed $offset): void {}
}

function viaSugar(Bag $b): void { echo (string) $b['q']; }
function viaCall(Bag $b): void  { echo (string) $b->offsetGet('q'); }
```

With `psalm.xml` at `errorLevel="1"` and `runTaintAnalysis="true"`:

**Expected:** both functions report.
**Actual:** only `viaCall` reports `TaintedHtml` and `TaintedTextWithQuotes`. `viaSugar` reports nothing.

```
ERROR Bag.php:13:40 TaintedHtml: Detected tainted HTML
ERROR Bag.php:13:40 TaintedTextWithQuotes: Detected tainted text with possible quotes
```

(Line 13 is `viaCall`. The `MissingPureAnnotation` errors that `errorLevel="1"` also emits are unrelated to this report.)

### Case 2: no source annotation, plain argument-to-return flow

The same edge is lost with no taint annotation anywhere, which shows the problem is structural rather than specific to source handling:

```php
<?php
/** @implements ArrayAccess<string, mixed> */
final class Passthru implements ArrayAccess
{
    #[Override] public function offsetGet(mixed $offset): mixed { return $offset; }
    #[Override] public function offsetExists(mixed $offset): bool { return true; }
    #[Override] public function offsetSet(mixed $offset, mixed $value): void {}
    #[Override] public function offsetUnset(mixed $offset): void {}
}

function sugarArg(Passthru $p): void { echo (string) $p[$_GET['x']]; }
function callArg(Passthru $p): void  { echo (string) $p->offsetGet($_GET['x']); }
```

**Actual:** only `callArg` reports `TaintedHtml` / `TaintedTextWithQuotes`. `sugarArg` reports nothing, even though the tainted value is passed in and returned straight back out.

### Mechanism

`ArrayFetchAnalyzer::handleArrayAccessOnNamedObject()` resolves the array-access form by synthesizing a `VirtualMethodCall` to `offsetGet()`. It saves `$statements_analyzer->node_data`, replaces it with a **clone**, runs `MethodCallAnalyzer::analyze()` against that clone, copies back only the resulting **type**, then restores the original node data:

```php
$old_node_data = $statements_analyzer->node_data;
$statements_analyzer->node_data = clone $statements_analyzer->node_data;

$fake_get_method_call = new VirtualMethodCall(
    $stmt->var,
    new VirtualIdentifier('offsetGet', $stmt->var->getAttributes()),
    [new VirtualArg($stmt->dim)],
);

MethodCallAnalyzer::analyze($statements_analyzer, $fake_get_method_call, $context);

$call_array_access_type =
    $statements_analyzer->node_data->getType($fake_get_method_call) ?? Type::getMixed();

$statements_analyzer->node_data = $old_node_data;
```

Everything the method-call analysis recorded about that synthesized call is confined to the discarded clone, so nothing links the outer `ArrayDimFetch` expression to the flow established inside. Plain type inference is unaffected, because the type is explicitly copied back out before the clone is dropped. That is precisely why this stays invisible without a taint test.

The same clone-and-discard shape appears in the sibling `offsetSet` and `DOMNodeList::item` branches in that method, so the write direction is worth checking as well.

Observed in `ArrayFetchAnalyzer.php` around lines 1904 to 1931 on the version below. Please confirm the current range rather than trusting those numbers.

### Suggested fix

Either:

1. Link the synthesized call's return node to the outer expression's node before discarding the clone, so the `ArrayDimFetch` inherits the parent nodes, or
2. Analyze the `VirtualMethodCall` against the live node-data set instead of a clone, if the clone is only there to keep unrelated inferred types from leaking.

### Environment

Verified on **Psalm 7.0.0-beta19** (`7e751c06a756fa64dc4c759c09fe4a173afcb433`), PHP 8.4, **with no plugins loaded**, using a minimal `psalm.xml` that registers none. Calling this out explicitly because the behaviour was found while working on a Psalm plugin, and it is a reasonable first suspicion that a plugin caused it. It does not.

### Related

* #11530 is the same family for `Iterator::current()` reached through `foreach`, where the virtual `current()` call likewise fails to propagate taint to the loop variable. A fix that makes synthesized calls contribute to the data flow graph would likely address both.
* psalm/psalm-plugin-laravel#1304 hit this exact shape for `Illuminate\Http\Request` and was closed as not planned there, on the grounds that the correct layer for the fix is Psalm core rather than a plugin.
